src/docs/design-patterns/cqrs-pattern.md

```diff
@@ -0,0 +1,50 @@
+---
+title: "CQRS - Command Query Responsibility Segregation"
+date: "2023-08-11"
+description: CQRS is a design pattern used in software architecture to address the complexity and performance issues that can arise in systems handling both read (query) and write (command) operations.
+featuredImage: "./images/cqrs-pattern.png"
+---
+
+**CQRS** - Command Query Responsibility Segregation - is a design pattern used in software architecture to address the complexity and performance issues that can arise in systems handling both read (**query**) and write (**command**) operations. CQRS suggests segregating the data models and operations for reads and writes into separate components, optimizing each for its specific use case.
+
+In a traditional monolithic architecture, the same data model and database schema are often used for both reading and writing data. However, as an application grows in complexity, the read and write patterns might have different requirements and performance characteristics. CQRS helps to tackle this by decoupling the data storage and retrieval mechanisms.
+
+## How CQRS Works
+
+These are some key features of CQRS and how it works.
+
+### Separation of Models
+
+Instead of using a single data model for both reading and writing, CQRS recommends having separate models. The command model is optimized for handling write operations, ensuring data consistency and integrity. The query model is designed for efficient data retrieval and is optimized for read operations.
+
+### Command and Query Paths
+
+The application's code is divided into two distinct paths: the Command Path and the Query Path. The Command Path handles write operations, such as creating, updating, or deleting data. The Query Path handles read operations and is optimized for querying and displaying data to users.
+
+![CQRS flowchart - the Query Path represents the user getting data from a read data store. The Command Path represents the user changing data in a writeable data store.](images/cqrs-flow.png)
+
+### Event Sourcing
+
+In many CQRS implementations, event sourcing is used to capture all changes to the application state as a sequence of events. These events represent the transitions in the system's data. By replaying these events, you can reconstruct the state of the system at any point in time. This approach provides a reliable audit trail and can facilitate debugging and troubleshooting.
+
+
+### Asynchronous Processing
+
+CQRS often involves asynchronous communication between the Command and Query sides. Commands are processed by the Command Model and can generate events. These events are then propagated to the Query Model, which updates its data store asynchronously. This separation of processing can improve scalability and responsiveness.
```

Comment:
Maybe worth mentioning explicitly that commands can be dispatched and their work done outside the process that initiated the command. A web API endpoint or server POST can return immediately, having dispatched the command, before its processing has completed (or even begun).